File: source/com/barrybecker4/puzzle/twopails/model/PourOperation.java
/** Copyright by Barry G. Becker, 20013. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.puzzle.twopails.model;


import com.barrybecker4.puzzle.common.model.Move;

/**
 * There are 6 legal pour operations:
 * 1) fill first container to top
 * 2) fill second container to top
 * 3) empty first container
 * 4) empty second container
 * 5) pour everything in first container to second.
 * 6) pour everything in second container to first.
 * Immutable.
 *@author Barry Becker
 */
public final class PourOperation implements Move {

    public enum Action {FILL, EMPTY, TRANSFER}
    public enum Container {FIRST, SECOND}

    private Action action;
    private Container container;

    /**
     * create a pour operation representing a state transition .
     */
    PourOperation(Action action, Container container) {
        this.action = action;
        this.container = container;
    }

    Action getAction() {
        return action;
    }

    Container getContainer() {
        return container;
    }

    /**
     * @return the reverse of this operation
     */
    PourOperation reverse() {
        Action action = getAction();
        Container container = getContainer();
        switch (getAction()) {
            case FILL : action = Action.EMPTY; break;
            case EMPTY: action = Action.FILL; break;
            case TRANSFER: container = getContainer() == Container.FIRST ? Container.SECOND : Container.FIRST;
        }
        return new PourOperation(action, container);
    }

    @Override
    public String toString() {
        StringBuilder s = new StringBuilder();
        s.append(action).append(" ").append(container);
        return s.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        PourOperation that = (PourOperation) o;
        return action == that.action && container == that.container;
    }

    @Override
    public int hashCode() {
        int result = action.hashCode();
        result = 31 * result + container.hashCode();
        return result;
    }
}

File: scala-source/com/barrybecker4/puzzle/twopails/model/PourOperation.scala
// Copyright by Barry G. Becker, 20017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.twopails.model

import com.barrybecker4.puzzle.common.model.Move
import PourOperation.{Action, Container}

/**
  * There are 6 legal pour operations:
  * 1) fill first container to top
  * 2) fill second container to top
  * 3) empty first container
  * 4) empty second container
  * 5) pour everything in first container to second.
  * 6) pour everything in second container to first.
  * Immutable.
  *
  * @author Barry Becker
  */
object PourOperation {

  enum Action:
    case FILL, EMPTY, TRANSFER

  enum Container:
    case FIRST, SECOND
}

case class PourOperation(action: Action, container: Container) extends Move {

  /** The reverse of this operation (undo step). */
  private[model] def reverse: PourOperation =
    action match
      case Action.FILL => PourOperation(Action.EMPTY, container)
      case Action.EMPTY => PourOperation(Action.FILL, container)
      case Action.TRANSFER =>
        val src = if container == Container.FIRST then Container.SECOND else Container.FIRST
        PourOperation(Action.TRANSFER, src)

  override def toString: String =
    action.toString + " " + container.toString
}
